Add yargs for flags, and several follow-up commits for new flags

nfuller opened the pull request:
The first adds package-lock.json, which I confess I don't really understand as npm is new to me.

I've split the commits to make it easier to review each individually.

evansiroky commented on README.md:
````diff
@@ -51,7 +51,7 @@ node --max-old-space-size=8192 index.js
 
 ```shell
-node --max-old-space-size=8192 index.js --filtered-zones "America/New_York,America/Chicago"
+node --max-old-space-size=8192 index.js --included_zones America/New_York America/Chicago
````
Could you also add some examples and usages for the new arguments in the script?

nfuller replied:
done

evansiroky commented on index.js:
```diff
@@ -693,46 +745,77 @@ const autoScript = {
-    overallProgress.beginTask('Zipping geojson')
-    exec('zip dist/timezones.geojson.zip dist/combined.json', cb)
+    if (argv.skip_zip) {
+      overallProgress.beginTask('Skipping zip')
```
The callback `cb` needs to be called in this block in order to indicate that this task has completed successfully. Also, one small code style guide is that it could be cleaner to include a return statement in this block and then the rest of the method won't be so indented as much. ie:

```js
if (argv.skip_zip) {
  overallProgress.beginTask('Skipping zip')
  return cb()
}
overallProgress.beginTask('Zipping geojson')
...
```

nfuller replied:
done.

evansiroky commented on index.js:
```diff
@@ -693,46 +745,77 @@ const autoScript = {
   zipGeoJsonWithOceans: ['mergeZones', function (results, cb) {
-    overallProgress.beginTask('Zipping geojson with oceans')
-    exec('zip dist/timezones-with-oceans.geojson.zip dist/combined-with-oceans.json', cb)
+    if (argv.skip_zip) {
+      overallProgress.beginTask('Skipping with oceans zip')
+    } else {
+      overallProgress.beginTask('Zipping geojson with oceans')
```
The callback `cb` needs to be called in this block in order to indicate that this task has completed successfully. Also, one small code style guide is that it could be cleaner to include a return statement in this block and then the rest of the method won't be so indented as much. ie:

```js
if (argv.skip_zip) {
  overallProgress.beginTask('Skipping with oceans zip')
  return cb()
}
overallProgress.beginTask('Zipping geojson with oceans')
...
```

nfuller replied:
Done in commit 5

evansiroky commented on index.js:
```diff
@@ -693,46 +745,77 @@ const autoScript = {
-        exec('zip dist/timezones.shapefile.zip dist/combined-shapefile.*', cb)
-      }
-    )
+    if (argv.skip_shapefile) {
+      overallProgress.beginTask('Skipping shapefile creation')
+    } else {
+      overallProgress.beginTask('Converting from geojson to shapefile')
```
The callback `cb` needs to be called in this block in order to indicate that this task has completed successfully. Also, one small code style guide is that it could be cleaner to include a return statement in this block and then the rest of the method won't be so indented as much. ie:

```js
if (argv.skip_shapefile) {
  overallProgress.beginTask('Skipping shapefile creation')
  return cb()
}
overallProgress.beginTask('Converting from geojson to shapefile')
...
```

nfuller replied:
Done in commit 5

evansiroky commented on index.js:
```diff
@@ -693,46 +745,77 @@ const autoScript = {
-        exec('zip dist/timezones-with-oceans.shapefile.zip dist/combined-shapefile-with-oceans.*', cb)
-      }
-    )
+    if (argv.skip_shapefile) {
+      overallProgress.beginTask('Skipping with oceans shapefile creation')
+    } else {
+      overallProgress.beginTask('Converting from geojson with oceans to shapefile')
+      let shapeFileGlob = argv.dist_dir + '/combined-shapefile-with-oceans.*'
```
The callback `cb` needs to be called in this block in order to indicate that this task has completed successfully. Also, one small code style guide is that it could be cleaner to include a return statement in this block and then the rest of the method won't be so indented as much. ie:

```js
if (argv.skip_shapefile) {
  overallProgress.beginTask('Skipping with oceans shapefile creation')
  return cb()
}
overallProgress.beginTask('Converting from geojson with oceans to shapefile')
...
```

nfuller replied:
Done in commit 5

nfuller replied:
Done in commit 5

nfuller commented on README.md:
```diff
@@ -54,9 +54,35 @@ node --max-old-space-size=8192 index.js
 ### What the script does
 
-There are three config files that describe the boundary building process.  The `osmBoundarySources.json` file lists all of the needed boundaries to extract via queries to the Overpass API.  The `timezones.json` file lists all of the timezones and various operations to perform to build the boundaries.  The `expectedZoneOverlaps.json` file lists all timezones that are allowed to overlap each other  ...
+There are three config files that describe the boundary building process.  The `osmBoundarySources.json` file lists all of the needed boundaries to extract via queries to the Overpass API.  The `timezones.json` file lists all of the timezones and various operations to perform to build the boundaries.  The `expectedZoneOverlaps.json` file lists all timezones that are allowed to overlap each other  ...
```
Just a whitespace removal I noticed.

nfuller commented on CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+## Pending
```
Not sure what you wanted to put here so I just put "Pending".

evansiroky commented on index.js:
```diff
@@ -11,23 +11,72 @@ var asynclib = require('async')
 var jsts = require('jsts')
 var rimraf = require('rimraf')
 var overpass = require('query-overpass')
+var yargs = require('yargs')
```
yargs is a new dependency for this library and must be added to the list of dependencies in `package.json`.

evansiroky replied:
It's cool to check in the package-lock.json file too.

nfuller replied:
I added and updated with "npm update". LMK if that's not correct.

evansiroky commented on index.js:
```diff
@@ -276,7 +325,7 @@ var getTzDistFilename = function (tzid) {
   var geoJson
   if (source.source === 'overpass') {
-    geoJson = require('./downloads/' + source.id + '.json')
+    geoJson = require(argv.downloads_dir + '/' + source.id + '.json')
```
If I try putting in the path `downloads-test` without the `./` prefix this will fail. If this is changed to `path.resolve(argv.downloads_dir)`, then it will work. It might be a good idea to wrap each of these arguments that are folders with a path.resolve to help with this.

nfuller replied:
I've done as you suggested.

evansiroky commented on index.js:
```diff
@@ -677,7 +729,7 @@ const autoScript = {
     overallProgress.beginTask('Validating timezone boundaries')
     loadDistZonesIntoMemory()
-    if (process.argv.indexOf('no-validation') > -1) {
+    if (argv.no_validation) {
```
I can't comment on the line directly, but the `zipInputData` function still has some hardcoded values that should probably be changed to the folders received from yargs parsing the cli input.

nfuller replied:
Done.

nfuller commented:
I've fixed the lint errors.

For the package-lock.json / package.json changes, I'm not very clear how this is supposed to work (but it looks like I'm not the only one out there from a quick search). I've tried to make the minimal changes for this PR. I reverted to the version of those files before my PR, and then just did what I think are the minimal steps required to record the new explicit dependency on yargs.

[I went on a brief diversion when reverted to the versions before my changes, deleted node_modules, and found I could still invoke "node index.js --help" successfully. Eventually I tracked that down to yargs already being a transitive dependency via some other deps.]

I resisted the temptation to run "npm audit fix", but you might want to look at that as a follow up.

Given there are 8 commits in this PR now, I will not be offended if you want to collapse them into a single commit - I kept them separate so you could see the progression but it means your history will be more messy than necessary.

evansiroky reviewed (approved):
Thanks a bunch! It is great to have a proper CLI helper in this script.
